packages/adblocker/src/resources.ts

```diff
@@ -36,129 +65,212 @@ export interface Resource {
+        this.redirectDetails.set(name, {
+          body: redirect.content,
+          contentType: redirect.contentType,
+          dataUrl,
+        });
+      }
+    }
+    for (const scriptlet of scriptlets) {
+      for (const name of scriptlet.names) {
+        this.scriptlets.set(name, scriptlet);
+      }
+    }
   }
 
-  public getResource(name: string): Resource & { dataUrl: string } {
-    const { body, contentType } = this.resources.get(name) || getResourceForMime(name);
+  private generateDataURL(redirect: RedirectResource): string {
+    if (redirect.encoding === 'base64') {
+      return `data:${redirect.contentType};base64,${redirect.content}`;
+    }
+
+    return `data:${redirect.contentType},${redirect.content}`;
+  }
 
-    let dataUrl;
-    if (contentType.indexOf(';') !== -1) {
-      dataUrl = `data:${contentType},${body}`;
-    } else {
-      dataUrl = `data:${contentType};base64,${btoaPolyfill(body)}`;
+  public getRedirect(name: string): RedirectDetail {
+    const detail = this.redirectDetails.get(name);
+    if (detail !== undefined) {
+      return detail;
     }
 
-    return { body, contentType, dataUrl };
+    // Fallback to sample resources
+    const sample = getResourceForMime(name);
+    const dataUrl = sample.contentType.includes(';')
+      ? `data:${sample.contentType};base64,${btoaPolyfill(sample.body)}`
+      : `data:${sample.contentType},${sample.body}`;
+
+    return {
+      body: sample.body,
+      contentType: sample.contentType,
+      dataUrl,
+    };
   }
 
   public getSerializedSize(): number {
-    let estimatedSize = sizeOfASCII(this.checksum) + 2 * sizeOfByte(); // resources.size
+    let estimatedSize = sizeOfASCII(this.checksum);
 
-    this.resources.forEach(({ contentType, body }, name) => {
-      estimatedSize += sizeOfASCII(name) + sizeOfASCII(contentType) + sizeOfUTF8(body);
-    });
+    for (const redirect of new Set(this.redirects.values())) {
+      estimatedSize += redirect.names.reduce((state, name) => state + sizeOfASCII(name), 4);
+
+      estimatedSize += sizeOfUTF8(redirect.content);
+      estimatedSize += sizeOfASCII(redirect.contentType);
+      estimatedSize += 1;
+    }
+
+    for (const scriptlet of new Set(this.scriptlets.values())) {
+      estimatedSize += scriptlet.names.reduce((state, name) => state + sizeOfASCII(name), 4);
+
+      estimatedSize += sizeOfUTF8(scriptlet.content);
+      estimatedSize += scriptlet.dependencies.reduce(
+        (state, name) => state + sizeOfASCII(name),
+        4,
+      );
+      estimatedSize += 2;
+    }
 
     return estimatedSize;
   }
 
   public serialize(buffer: StaticDataView): void {
-    // Serialize `checksum`
     buffer.pushASCII(this.checksum);
 
-    // Serialize `resources`
-    buffer.pushUint16(this.resources.size);
-    this.resources.forEach(({ contentType, body }, name) => {
-      buffer.pushASCII(name);
-      buffer.pushASCII(contentType);
-      buffer.pushUTF8(body);
-    });
```

Comment:
do we need to change the way we handle redirect resources? perhaps we can continue to use all that old code?